File: Android/React-Native/src/api/youtube.ts
import { backendClient } from "./backendClient";
import { API_ENDPOINTS } from "./config";
import { VideoInfo, SearchResult } from "../models/videoModel";

/**
 * Backend response types
 */
export type BackendVideoInfo = {
  videoId: string;
  title: string;
  thumbnailUrl: string;
  duration: number;
  channel: string;
  viewCount: number;
  uploadDate?: string;
  url: string;
};

export type BackendSearchResult = {
  videos: BackendVideoInfo[];
  query: string;
  page: number;
  hasMore: boolean;
};

export type InitiateDownloadResponse = {
  downloadId: string;
  videoId: string;
  downloadUrl: string;
  fileName: string;
  fileSize?: number;
  format: string;
  message: string;
};

export type DownloadStatusResponse = {
  downloadId: string;
  status: "pending" | "downloading" | "completed" | "failed" | "cancelled";
  progress: number;
  downloadedBytes: number;
  totalBytes: number;
  error?: string;
};

/**
 * YouTube API functions using backend server
 */
export class YouTubeAPI {
  /**
   * Search for videos
   */
  static async searchVideos(query: string, maxResults: number = 20): Promise<SearchResult> {
    try {
      console.log(`[YouTube API] Searching for: ${query}`);
      
      const response = await backendClient.post<BackendSearchResult>(API_ENDPOINTS.SEARCH, {
        query,
        maxResults,
      });

      // Convert backend format to app format
      const videos: VideoInfo[] = response.videos.map((video) => ({
        videoId: video.videoId,
        title: video.title,
        thumbnailUrl: video.thumbnailUrl,
        duration: video.duration,
        channel: video.channel,
        viewCount: video.viewCount,
        uploadDate: video.uploadDate,
        url: video.url,
        isPlaylist: false,
      }));

      return {
        videos,
        query: response.query,
        page: response.page || 1,
        hasMore: response.hasMore || false,
      };
    } catch (error) {
      console.error("[YouTube API] Search failed:", error);
      throw error;
    }
  }

  /**
   * Get video information
   */
  static async getVideoInfo(url: string): Promise<VideoInfo | null> {
    try {
      console.log(`[YouTube API] Fetching video info for: ${url}`);
      
      const response = await backendClient.post<BackendVideoInfo>(API_ENDPOINTS.VIDEO_INFO, {
        url,
      });

      return {
        videoId: response.videoId,
        title: response.title,
        thumbnailUrl: response.thumbnailUrl,
        duration: response.duration,
        channel: response.channel,
        viewCount: response.viewCount,
        uploadDate: response.uploadDate,
        url: response.url,
        isPlaylist: false,
      };
    } catch (error) {
      console.error("[YouTube API] Get video info failed:", error);
      return null;
    }
  }

  /**
   * Initiate video download
   * Returns download URL and metadata from backend
   */
  static async initiateDownload(
    url: string,
    formatType: "mp4" | "mp3"
  ): Promise<InitiateDownloadResponse> {
    try {
      console.log(`[YouTube API] Initiating download: ${url} (${formatType})`);
      
      const response = await backendClient.post<InitiateDownloadResponse>(API_ENDPOINTS.DOWNLOAD, {
        url,
        format: formatType,
      });

      console.log(`[YouTube API] Download initiated: ${response.downloadId}`);
      return response;
    } catch (error) {
      console.error("[YouTube API] Initiate download failed:", error);
      throw error;
    }
  }

  /**
   * Get download status
   */
  static async getDownloadStatus(downloadId: string): Promise<DownloadStatusResponse> {
    try {
      const response = await backendClient.get<DownloadStatusResponse>(
        `${API_ENDPOINTS.DOWNLOAD_STATUS}/${downloadId}`
      );
      return response;
    } catch (error) {
      console.error("[YouTube API] Get download status failed:", error);
      throw error;
    }
  }

  /**
   * Cancel download
   */
  static async cancelDownload(downloadId: string): Promise<void> {
    try {
      console.log(`[YouTube API] Cancelling download: ${downloadId}`);
      
      await backendClient.post(API_ENDPOINTS.CANCEL_DOWNLOAD, {
        downloadId,
      });
      
      console.log(`[YouTube API] Download cancelled: ${downloadId}`);
    } catch (error) {
      console.error("[YouTube API] Cancel download failed:", error);
      throw error;
    }
  }
}

export default YouTubeAPI;

File: Android/Backend/app/routes/youtube.py
from fastapi import APIRouter, HTTPException, BackgroundTasks
from fastapi.responses import FileResponse
import os

from models.schemas import (
    SearchRequest,
    SearchResult,
    VideoInfoRequest,
    VideoInfo,
    DownloadRequest,
    DownloadResponse,
    DownloadStatus,
    CancelDownloadRequest,
)
from services.ytdlp_service import ytdlp_service

router = APIRouter()


@router.post("/search", response_model=SearchResult)
async def search_videos(request: SearchRequest):
    """
    Search for YouTube videos
    """
    try:
        videos_dict = await ytdlp_service.search_videos(request.query, request.maxResults)
        videos = [VideoInfo(**video) for video in videos_dict]
        
        return SearchResult(
            videos=videos,
            query=request.query,
            page=1,
            hasMore=len(videos) >= request.maxResults,
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Search failed: {str(e)}")


@router.post("/video-info", response_model=VideoInfo)
async def get_video_info(request: VideoInfoRequest):
    """
    Get information about a specific video
    """
    try:
        video_info = await ytdlp_service.get_video_info(request.url)
        
        if not video_info:
            raise HTTPException(status_code=404, detail="Video not found")
        
        return VideoInfo(**video_info)
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to get video info: {str(e)}")


@router.post("/download", response_model=DownloadResponse)
async def initiate_download(request: DownloadRequest, background_tasks: BackgroundTasks):
    """
    Initiate video download
    Returns download ID and URL for client to download the file
    """
    try:
        # Start download asynchronously
        download_info = await ytdlp_service.download_video(
            request.url,
            request.format,
        )
        
        return DownloadResponse(**download_info)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Download initiation failed: {str(e)}")


@router.get("/download-status/{download_id}", response_model=DownloadStatus)
async def get_download_status(download_id: str):
    """
    Get status of a download
    """
    status = ytdlp_service.get_download_status(download_id)
    
    if not status:
        raise HTTPException(status_code=404, detail="Download not found")
    
    return DownloadStatus(**status)


@router.get("/download-file/{download_id}")
async def download_file(download_id: str):
    """
    Download the completed file
    """
    filepath = ytdlp_service.get_download_file(download_id)
    
    if not filepath or not os.path.exists(filepath):
        raise HTTPException(status_code=404, detail="File not found or download not completed")
    
    filename = os.path.basename(filepath)
    
    return FileResponse(
        path=filepath,
        filename=filename,
        media_type="application/octet-stream",
    )


@router.post("/cancel-download")
async def cancel_download(request: CancelDownloadRequest):
    """
    Cancel an active download
    """
    success = ytdlp_service.cancel_download(request.downloadId)
    
    if not success:
        raise HTTPException(status_code=404, detail="Download not found")
    
    return {"message": "Download cancelled successfully", "downloadId": request.downloadId}
